File: internal/models/user_profile.go
package models

import (
	"github.com/google/uuid"
	"github.com/juevigrace/diva-server/internal/models/responses"
	"github.com/juevigrace/diva-server/storage"
)

type UserProfile struct {
	UserID    uuid.UUID
	FirstName string
	LastName  string
	BirthDate int64
	Alias     string
	Avatar    string
	Bio       string
	UpdatedAt int64
}

func (up *UserProfile) Response() *responses.UserProfileResponse {
	return &responses.UserProfileResponse{
		FirstName: up.FirstName,
		LastName:  up.LastName,
		BirthDate: up.BirthDate,
		Alias:     up.Alias,
		Avatar:    up.Avatar,
		Bio:       up.Bio,
	}
}

func (up *UserProfile) DBCreate(userID uuid.UUID) *storage.CreateUserProfileParams {
	return &storage.CreateUserProfileParams{
		UserID:    userID,
		FirstName: up.FirstName,
		LastName:  up.LastName,
		BirthDate: &up.BirthDate,
		Alias:     up.Alias,
		Bio:       up.Bio,
	}
}

func (up *UserProfile) DBUpdate(userID uuid.UUID) *storage.UpdateUserProfileParams {
	return &storage.UpdateUserProfileParams{
		UserID:    userID,
		FirstName: up.FirstName,
		LastName:  up.LastName,
		BirthDate: &up.BirthDate,
		Alias:     up.Alias,
		Bio:       up.Bio,
	}
}

func UserProfileFromDB(row *storage.DivaUserProfile) *UserProfile {
	birthDate := int64(0)
	if row.BirthDate != nil {
		birthDate = *row.BirthDate
	}
	return &UserProfile{
		UserID:    row.UserID,
		FirstName: row.FirstName,
		LastName:  row.LastName,
		BirthDate: birthDate,
		Alias:     row.Alias,
		Bio:       row.Bio,
		Avatar:    row.Avatar,
		UpdatedAt: row.UpdatedAt,
	}
}

File: client/packages/shared/app-lib/diva-lib-models-core/src/commonMain/kotlin/io/github/juevigrace/diva/lib/models/user/profile/UserProfile.kt
@file:OptIn(ExperimentalJsExport::class)
@file:DivaJsExport

package io.github.juevigrace.diva.lib.models.user.profile

import io.github.juevigrace.diva.core.DivaJsExport
import io.github.juevigrace.diva.core.None
import io.github.juevigrace.diva.core.Option
import io.github.juevigrace.diva.lib.models.api.user.profile.UserProfileResponse
import kotlin.js.ExperimentalJsExport
import kotlin.time.Instant
import kotlin.uuid.ExperimentalUuidApi
import kotlin.uuid.Uuid

@OptIn(ExperimentalUuidApi::class)
data class UserProfile(
    val userId: Uuid = Uuid.NIL,
    val firstName: String = "",
    val lastName: String = "",
    val birthDate: Option<Instant> = None,
    val phoneNumber: String = "",
    val alias: String = "",
    val avatar: String = "",
    val bio: String = "",
    val updatedAt: Option<Instant> = None,
) {
    companion object {
        fun fromResponse(response: UserProfileResponse): UserProfile {
            return UserProfile(
                firstName = response.firstName,
                lastName = response.lastName,
                birthDate = Option.of(Instant.fromEpochMilliseconds(response.birthDate)),
                phoneNumber = response.phoneNumber,
                alias = response.alias,
                avatar = response.avatar,
                bio = response.bio,
            )
        }
    }
}
